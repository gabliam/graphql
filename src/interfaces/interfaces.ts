import { inversifyInterfaces } from '@gabliam/core';
import * as GraphiQL from 'graphql-server-module-graphiql';


export type listControllers = inversifyInterfaces.ServiceIdentifier<any>[];

export type resolverType = 'Query' | 'Mutation' | 'Subscription';

export interface GraphiqlOptions {
  subscriptionsEndpoint?: string;
  query?: string;
  variables?: Object;
  operationName?: string;
  result?: Object;
  passHeader?: string;
}

export interface GraphqlConfig {
  endpointUrl: string;

  endpointUrlGraphiql: string;

  graphiqlOptions: GraphiQL.GraphiQLData;

  graphqlFiles: string[] | undefined;
}
