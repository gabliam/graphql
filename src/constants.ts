/** Type in registry */
export const TYPE = {
  Controller: 'GraphqlController',
};

export const METADATA_KEY = {
  controller: '_controller',
  resolver: '_resolver',
};


export const GRAPHQL_PLUGIN_CONFIG = Symbol('GABLIAM/GRAPHQL_PLUGIN_CONFIG');

export const LIST_CONTROLLER_GRAPHQL = Symbol('LIST_CONTROLLER_GRAPHQL');

export const ROOT_SCHEMA = Symbol('ROOT_SCHEMA');

export const DEFAULT_END_POINT_URL = '/graphql';

export const DEFAULT_END_POINT_URL_GRAPHIQL = '/graphiql';
