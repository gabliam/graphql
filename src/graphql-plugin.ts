import {
  interfaces as coreInterfaces,
  inversifyInterfaces,
  Scan,
  Registry
} from '@gabliam/core';
import { MiddlewareConfig } from '@gabliam/express';
import { makeExecutableSchema } from 'graphql-tools';
import {
  TYPE,
  METADATA_KEY,
  GRAPHQL_PLUGIN_CONFIG,
  DEBUG_PATH
} from './constants';
import {
  ResolverMetadata,
  ControllerMetadata,
  GraphqlConfig
} from './interfaces';
import { IResolvers } from 'graphql-tools/dist/Interfaces';
import { graphqlExpress, graphiqlExpress } from 'graphql-server-express';
import * as bodyParser from 'body-parser';
import * as _ from 'lodash';
import * as fs from 'fs';
import * as d from 'debug';

const debug = d(DEBUG_PATH);

@Scan(__dirname)
export class GraphqlPlugin implements coreInterfaces.GabliamPlugin {
  bind(container: inversifyInterfaces.Container, registry: Registry) {
    registry.get(TYPE.Controller).map(({ id, target }) => {
      container
        .bind<any>(id)
        .to(target)
        .inSingletonScope();
      return id;
    });
  }

  build(container: inversifyInterfaces.Container, registry: Registry) {
    const middlewareConfig = container.get<MiddlewareConfig>(MiddlewareConfig);
    const graphqlPluginConfig = container.get<GraphqlConfig>(
      GRAPHQL_PLUGIN_CONFIG
    );
    const listdefinitions: string[] = [];

    if (graphqlPluginConfig.graphqlFiles) {
      listdefinitions.push(
        ...this.loadGraphqlFiles(...graphqlPluginConfig.graphqlFiles)
      );
    }

    const controllerIds = registry.get(TYPE.Controller);
    const resolverList: IResolvers[] = [];

    for (const { id: controllerId } of controllerIds) {
      const controller = container.get<object>(controllerId);

      const controllerMetadata: ControllerMetadata = Reflect.getOwnMetadata(
        METADATA_KEY.controller,
        controller.constructor
      );

      listdefinitions.push(...controllerMetadata.schema);
      listdefinitions.push(
        ...this.loadGraphqlFiles(...controllerMetadata.graphqlFiles)
      );

      const resolverMetadatas: ResolverMetadata[] = Reflect.getOwnMetadata(
        METADATA_KEY.resolver,
        controller.constructor
      );

      if (resolverMetadatas && controllerMetadata) {
        for (const resolverMetadata of resolverMetadatas) {
          if (resolverMetadata.schema) {
            listdefinitions.push(resolverMetadata.schema);
          }
          if (resolverMetadata.graphqlFile) {
            listdefinitions.push(
              ...this.loadGraphqlFiles(resolverMetadata.graphqlFile)
            );
          }

          resolverList.push(
            this.getResolver(container, controllerId, resolverMetadata)
          );
        }
      }
    }

    const queries = [];
    const mutations = [];
    const others = [];

    for (const definitions of listdefinitions) {
      const types = definitions
        .split('}')
        .map(s => s.replace(/^\s+|\s+$/g, '')) // remove space
        .filter(s => s !== '}' && s !== '') // remove usless
        .filter(s => s[0] !== '#') // remove commentaries lines
        .map(s => s + '\n}');

      for (let type of types) {
        if (/\s*type\s*Query\s*{\s*/g.test(type)) {
          if (type.slice(0, 6) !== 'extend') {
            // add extend
            type = `extend ${type}`;
          }
          queries.push(type);
        } else if (/\s*type\s*Mutation\s*{\s*/g.test(type)) {
          if (type.slice(0, 6) !== 'extend') {
            type = `extend ${type}`;
          }
          mutations.push(type);
        } else {
          others.push(type);
        }
      }
    }

    const typeDefs: string[] = [];
    // for first Type Query remove extend
    if (queries.length) {
      queries[0] = queries[0].slice(7);
      typeDefs.push(...queries);
    }

    if (mutations.length) {
      mutations[0] = mutations[0].slice(7);
      typeDefs.push(...mutations);
    }

    typeDefs.push(...others);

    const resolvers = _.merge<IResolvers>({}, ...resolverList);

    if (Object.keys(resolvers).length === 0) {
      return;
    }

    debug('typeDefs', typeDefs);
    const executableSchema = makeExecutableSchema({
      typeDefs,
      resolvers
    });
    middlewareConfig.addMiddleware({
      order: 50,
      instance: app => {
        debug('add graphql middleware to ExpressPlugin');
        app.use(bodyParser.urlencoded({ extended: true }));
        app.use(bodyParser.json());

        app.use(
          graphqlPluginConfig.endpointUrl,
          graphqlExpress(req => {
            let options = {};

            /* istanbul ignore if  */
            if ((<any>req).graphqlOptions) {
              options = (<any>req).graphqlOptions;
            }

            return {
              schema: executableSchema,
              ...options
            };
          })
        );

        app.use(
          graphqlPluginConfig.endpointUrlGraphiql,
          graphiqlExpress(graphqlPluginConfig.graphiqlOptions)
        );
      }
    });
  }

  private loadGraphqlFiles(...files: string[]) {
    return files.map(file => fs.readFileSync(file, 'UTF-8'));
  }

  private getResolver(
    container: inversifyInterfaces.Container,
    controllerId: any,
    resolverMetadata: ResolverMetadata
  ): IResolvers {
    const instance = container.get<any>(controllerId);
    const result: any = instance[resolverMetadata.key]();

    if (_.isFunction(result)) {
      // IResolverObject
      return _.set<IResolvers>(
        {},
        resolverMetadata.path,
        result.bind(instance)
      );
    } else {
      // GraphQLScalarType
      return _.set<IResolvers>({}, resolverMetadata.path, result);
    }
  }
}
