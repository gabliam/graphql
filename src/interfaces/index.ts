export * from './metadata';
export * from './interfaces';
