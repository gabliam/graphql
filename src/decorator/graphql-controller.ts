import { TYPE, METADATA_KEY } from '../constants';
import { injectable, register, inversifyInterfaces as interfaces } from '@gabliam/core';
import { ResolverMetadata, resolverType, ControllerMetadata } from '../interfaces';



/**
 * GraphqlController decorator
 *
 * Add a GraphqlController
 */
export function GraphqlController({
  schema = [],
  graphqlFiles = []
}: {
    schema?: string[]
    graphqlFiles?: string[]
  } = {
    schema: [],
    graphqlFiles: []
  }) {
  return (target: any) => {
    const id: interfaces.ServiceIdentifier<any> = target;
    injectable()(target);
    register(TYPE.Controller, { id, target })(target);
    const metadata: ControllerMetadata = { schema, graphqlFiles };
    Reflect.defineMetadata(METADATA_KEY.controller, metadata, target);
  }
}

export interface ResolverOptions {
  schema?: string;

  path?: string;

  graphqlFile?: string;
}


export function QueryResolver(options?: ResolverOptions) {
  return resolver('Query', options);
}
export function MutationResolver(options?: ResolverOptions) {
  return resolver('Mutation', options);
}
export function SubscriptionResolver(options?: ResolverOptions) {
  return resolver('Subscription', options);
}

export function Resolver(options?: ResolverOptions) {
  return resolver(null, options);
}

function resolver(type: resolverType | null, resolverOptions: ResolverOptions = {}) {
  return function (target: any, key: string, descriptor: PropertyDescriptor) {
    let resolverPath = '';
    if (resolverOptions.path) {
      resolverPath = resolverOptions.path
    } else {
      resolverPath = key;
    }

    if (type) {
      resolverPath = `${type}.${resolverPath}`;
    }

    const metadata: ResolverMetadata = {
      type,
      path: resolverPath,
      key,
      schema: resolverOptions.schema,
      graphqlFile: resolverOptions.graphqlFile
    };

    let metadataList: ResolverMetadata[] = [];

    if (!Reflect.hasOwnMetadata(METADATA_KEY.resolver, target.constructor)) {
      Reflect.defineMetadata(METADATA_KEY.resolver, metadataList, target.constructor);
    } else {
      metadataList = Reflect.getOwnMetadata(METADATA_KEY.resolver, target.constructor);
    }

    metadataList.push(metadata);
  };
}
