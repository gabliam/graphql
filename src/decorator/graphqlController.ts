import { TYPE, METADATA_KEY } from '../constants';
import { injectable, register, inversifyInterfaces as interfaces } from '@gabliam/core';
import { ResolverMetadata, resolverType, ControllerMetadata } from '../interfaces';


/**
 * GraphqlController decorator
 *
 * Add a GraphqlController
 * @param  {string} name? name of the GraphqlController
 */
export function GraphqlController(schema: string[]) {
  return (target: any) => {
    const id: interfaces.ServiceIdentifier<any> = target;
    injectable()(target);
    register(TYPE.Controller, { id, target })(target);
    const metadata: ControllerMetadata = { schema };
    Reflect.defineMetadata(METADATA_KEY.controller, metadata, target);
  }
}


export function QueryResolver(schema: string, path?: string) {
  return resolver('Query', schema);
}
export function MutationResolver(schema: string, path?: string) {
  return resolver('Mutation', schema);
}
export function SubscriptionResolver(schema: string, path?: string) {
  return resolver('Subscription', schema);
}

export function Resolver(path: string) {
  return resolver(null, null, path);
}

function resolver(type: resolverType | null, schema: string | null, path?: string) {
  return function (target: any, key: string, descriptor: PropertyDescriptor) {
    let resolverPath = '';
    if (path) {
      resolverPath = path
    } else {
      resolverPath = key;
    }

    if (type) {
      resolverPath = `${type}.${resolverPath}`;
    }

    const metadata: ResolverMetadata = { type, path: resolverPath, key, schema };
    let metadataList: ResolverMetadata[] = [];

    if (!Reflect.hasOwnMetadata(METADATA_KEY.resolver, target.constructor)) {
      Reflect.defineMetadata(METADATA_KEY.resolver, metadataList, target.constructor);
    } else {
      metadataList = Reflect.getOwnMetadata(METADATA_KEY.resolver, target.constructor);
    }

    metadataList.push(metadata);
  };
}
