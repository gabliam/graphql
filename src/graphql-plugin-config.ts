import { Value, Bean, PluginConfig } from '@gabliam/core';
import {
  DEFAULT_END_POINT_URL,
  DEFAULT_END_POINT_URL_GRAPHIQL,
  GRAPHQL_PLUGIN_CONFIG,
  DEBUG_PATH
} from './constants';
import * as Joi from 'joi';
import { GraphiqlOptions, GraphqlConfig } from './interfaces';
import * as d from 'debug';

const debug = d(`${DEBUG_PATH}:GraphqlPluginConfig`);

export const GraphiqlOptionsValidator = Joi.object().keys({
  subscriptionsEndpoint: Joi.string(),
  query: Joi.string(),
  variables: Joi.object(),
  operationName: Joi.number(),
  result: Joi.object(),
  passHeader: Joi.string()
});

@PluginConfig()
export class GraphqlPluginConfig {
  @Value('application.graphql.endpointUrl', Joi.string())
  private endpointUrl: string = DEFAULT_END_POINT_URL;

  @Value('application.graphql.graphqlFiles', Joi.array().items(Joi.string()))
  private graphqlFiles: string[] | undefined;

  @Value('application.graphiql.enabled', Joi.boolean())
  private graphiqlEnabled = true;

  @Value('application.graphiql.endpointUrl', Joi.string())
  private endpointUrlGraphiql: string = DEFAULT_END_POINT_URL_GRAPHIQL;

  @Value('application.graphiql.options', GraphiqlOptionsValidator)
  private graphiqlOptions: GraphiqlOptions = {};

  @Bean(GRAPHQL_PLUGIN_CONFIG)
  creatreConfig(): GraphqlConfig {
    const graphqlConfig: GraphqlConfig = {
      endpointUrl: this.endpointUrl,
      endpointUrlGraphiql: this.endpointUrlGraphiql,
      graphqlFiles: this.graphqlFiles,
      graphiqlEnabled: this.graphiqlEnabled,
      graphiqlOptions: {
        endpointURL: this.endpointUrl,
        ...this.graphiqlOptions
      }
    };

    debug('GraphqlConfig', graphqlConfig);

    return graphqlConfig;
  }
}
