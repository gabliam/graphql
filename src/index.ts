import * as interfaces from './interfaces';

import { GraphqlPlugin } from './graphql-plugin';

export * from './decorator';
export * from 'graphql-server-express';
export * from 'graphql-tools';
export * from 'graphql';

export {
  interfaces,
  GraphqlPlugin as default
};
