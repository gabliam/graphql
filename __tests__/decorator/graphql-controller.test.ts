// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { GraphqlController } from '../../src/decorator';
import { METADATA_KEY } from '../../src/constants';
import { ControllerMetadata } from '../../src/interfaces';
import { PhotoSchema, HeroSchema } from '../fixtures/schemas/schema';
import * as path from 'path';

test('should add GraphqlController metadata to a class when decorated with @GraphqlController()', () => {
  @GraphqlController()
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: absolute})', () => {
  @GraphqlController({
    graphqlFiles: [
      `${__dirname}/photo/hero.gql`,
      `${__dirname}/photo/photo.gql`
    ]
  })
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative})', () => {
  @GraphqlController({
    graphqlFiles: [`./photo/hero.gql`, `./photo/photo.gql`]
  })
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative, pwd})', () => {
  @GraphqlController({
    graphqlFiles: [`./photo/hero.gql`, `./photo/photo.gql`],
    pwd: path.resolve(__dirname, '..')
  })
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({schema})', () => {
  @GraphqlController({
    schema: [PhotoSchema, HeroSchema]
  })
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles, schema})', () => {
  @GraphqlController({
    graphqlFiles: [`${__dirname}/photo/photo.gql`],
    schema: [HeroSchema]
  })
  class TestBean {}

  const controllerMetadata: ControllerMetadata = Reflect.getMetadata(
    METADATA_KEY.controller,
    TestBean
  );

  expect(controllerMetadata).toMatchSnapshot();
});
