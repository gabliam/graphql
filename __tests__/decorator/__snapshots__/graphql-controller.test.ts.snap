// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController() 1`] = `
Object {
  "graphqlFiles": Array [],
  "schema": Array [],
}
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles, schema}) 1`] = `
Object {
  "graphqlFiles": Array [
    "<PROJECT_ROOT>/__tests__/decorator/photo/photo.gql",
  ],
  "schema": Array [
    "
type Hero {
  id: Int!

  name: String!

  power: String!

  amountPeopleSaved: Int!
}

input HeroInput {
  name: String!

  power: String!

  amountPeopleSaved: Int!
}

type HeroPage {
  items: [Hero]
  totalCount: Int
}

type Query {
  heroes: [Hero]
  getPageOfHeroes(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): HeroPage
}

type Mutation {
  submitHero(heroInput: HeroInput!): Hero
}
",
  ],
}
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: absolute}) 1`] = `
Object {
  "graphqlFiles": Array [
    "<PROJECT_ROOT>/__tests__/decorator/photo/hero.gql",
    "<PROJECT_ROOT>/__tests__/decorator/photo/photo.gql",
  ],
  "schema": Array [],
}
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative, pwd}) 1`] = `
Object {
  "graphqlFiles": Array [
    "<PROJECT_ROOT>/__tests__/photo/hero.gql",
    "<PROJECT_ROOT>/__tests__/photo/photo.gql",
  ],
  "schema": Array [],
}
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative}) 1`] = `
Object {
  "graphqlFiles": Array [
    "<PROJECT_ROOT>/__tests__/decorator/photo/hero.gql",
    "<PROJECT_ROOT>/__tests__/decorator/photo/photo.gql",
  ],
  "schema": Array [],
}
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({schema}) 1`] = `
Object {
  "graphqlFiles": Array [],
  "schema": Array [
    "
type Photo {
  id: Int!

  name: String!

  description: String!

  fileName: String!

  views: Int!

  isPublished: Boolean!
}

input PhotoInput {
  name: String!

  description: String!

  fileName: String!

  views: Int!

  isPublished: Boolean!
}

type PhotoPage {
  items: [Photo]
  totalCount: Int
}

type Query {
  photos: [Photo]
  getPageOfPhotos(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): PhotoPage
}

type Mutation {
  submitPhoto(photoInput: PhotoInput!): Photo
}
",
    "
type Hero {
  id: Int!

  name: String!

  power: String!

  amountPeopleSaved: Int!
}

input HeroInput {
  name: String!

  power: String!

  amountPeopleSaved: Int!
}

type HeroPage {
  items: [Hero]
  totalCount: Int
}

type Query {
  heroes: [Hero]
  getPageOfHeroes(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): HeroPage
}

type Mutation {
  submitHero(heroInput: HeroInput!): Hero
}
",
  ],
}
`;
